File: java/hard/q28.java
/*
Challenge: The Bank Robbery (Concurrency)

Scenario:
A shared bank account starts with $1000, and multiple threads represent customers.
Race conditions allow the balance to drop below zero.

Problem: they are not customers, but are robbers.

Debugging Tasks:
Mutually Exclude all shared variables from the threads using them.
*/


class BankAccount {
    private int balance;

    public BankAccount(int initialBalance) {
        this.balance = initialBalance;
    }

    public void withdraw(int amount) {
        if (balance >= amount) {
            System.out.println("Processing withdrawal of " + amount + "...");
            balance -= amount;
            System.out.println("Withdrawal successful. Remaining balance: " + balance);
        } else {
            System.out.println("Withdrawal of " + amount + " denied. Insufficient funds.");
        }
    }

    public int getBalance() {
        return balance;
    }
}

class Customer implements Runnable {
    private BankAccount account;
    private int withdrawalAmount;

    public Customer(BankAccount account, int withdrawalAmount) {
        this.account = account;
        this.withdrawalAmount = withdrawalAmount;
    }

    @Override
    public void run() {
        for (int i = 0; i < 5; i++) {
            account.withdraw(withdrawalAmount);
        }
    }
}

public class BankRobbery {
    public static void main(String[] args) {
        BankAccount account = new BankAccount(1000);

        Thread customer1 = new Thread(new Customer(account, 400));
        Thread customer2 = new Thread(new Customer(account, 400));
        Thread customer3 = new Thread(new Customer(account, 400));

        customer1.start();
        customer2.start();
        customer3.start();

        try {
            customer1.join();
            customer2.join();
            customer3.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        System.out.println("Final balance: " + account.getBalance());
    }
}

File: python/hard/q28.py
"""
Challenge: The Bank Robbery (Concurrency)

Scenario:
A shared bank account starts with $1000, and multiple threads represent customers.
Race conditions allow the balance to drop below zero.

Problem: they are not customers, but are robbers.

Debugging Tasks:
Mutually Exclude all shared variables from the threads using them.
"""

import threading

class BankAccount:
    def __init__(self, initial_balance):
        self.balance = initial_balance

    def withdraw(self, amount):
        if self.balance >= amount:
            print(f"Processing withdrawal of {amount}...")
            self.balance -= amount
            print(f"Withdrawal successful. Remaining balance: {self.balance}")
        else:
            print(f"Withdrawal of {amount} denied. Insufficient funds.")

    def get_balance(self):
        return self.balance

class Customer(threading.Thread):
    def __init__(self, account, withdrawal_amount):
        threading.Thread.__init__(self)
        self.account = account
        self.withdrawal_amount = withdrawal_amount

    def run(self):
        for _ in range(5):
            self.account.withdraw(self.withdrawal_amount)

if __name__ == "__main__":
    account = BankAccount(1000)

    customer1 = Customer(account, 400)
    customer2 = Customer(account, 400)
    customer3 = Customer(account, 400)

    customer1.start()
    customer2.start()
    customer3.start()

    customer1.join()
    customer2.join()
    customer3.join()

    print(f"Final balance: {account.get_balance()}")
